testsuite/src/main/java/io/netty/testsuite/transport/socket/SocketConnectTest.java

```diff
@@ -110,8 +125,109 @@ public void channelInactive(ChannelHandlerContext ctx) throws Exception {
         }
     }
 
+    @Test(timeout = 3000)
+    public void testWriteWithFastOpenBeforeConnect() throws Throwable {
+        run();
+    }
+
+    public void testWriteWithFastOpenBeforeConnect(ServerBootstrap sb, Bootstrap cb) throws Throwable {
+        enableTcpFastOpen(sb, cb);
+
+        sb.childHandler(new ChannelInitializer<SocketChannel>() {
+            @Override
+            protected void initChannel(SocketChannel ch) throws Exception {
+                ch.pipeline().addLast(new EchoServerHandler());
+            }
+        });
+
+        Channel sc = sb.bind().sync().channel();
+        connectAndVerifyDataTransfer(cb, sc);
+        connectAndVerifyDataTransfer(cb, sc);
+    }
+
+    private static void connectAndVerifyDataTransfer(Bootstrap cb, Channel sc)
+            throws InterruptedException {
+        BufferingClientHandler handler = new BufferingClientHandler();
+        cb.handler(handler);
+        ChannelFuture register = cb.register();
+        Channel channel = register.sync().channel();
+        ChannelFuture write = channel.write(writeAscii(DEFAULT, "[fastopen]"));
+        SocketAddress remoteAddress = sc.localAddress();
+        ChannelFuture connectFuture = channel.connect(remoteAddress);
+        Channel cc = connectFuture.sync().channel();
+        cc.writeAndFlush(writeAscii(DEFAULT, "[normal data]")).sync();
+        write.sync();
+        String expectedString = "[fastopen][normal data]";
+        String result = handler.collectBuffer(expectedString.getBytes(US_ASCII).length);
+        cc.disconnect().sync();
+        assertEquals(expectedString, result);
+    }
+
+    protected void enableTcpFastOpen(ServerBootstrap sb, Bootstrap cb) {
+        throw new AssumptionViolatedException(
+                "Support for testing TCP_FASTOPEN not enabled for " + StringUtil.simpleClassName(this));
+    }
+
     private static void assertLocalAddress(InetSocketAddress address) {
         assertTrue(address.getPort() > 0);
         assertFalse(address.getAddress().isAnyLocalAddress());
     }
+
+    private static class BufferingClientHandler extends ChannelInboundHandlerAdapter {
+        private final Semaphore semaphore = new Semaphore(0);
+        private final StringBuffer stringBuffer = new StringBuffer();
+
+        @Override
+        public void channelActive(ChannelHandlerContext ctx) throws Exception {
+            ctx.read();
+            super.channelActive(ctx);
+        }
+
+        @Override
+        public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception {
+            if (msg instanceof ByteBuf) {
+                ByteBuf buf = (ByteBuf) msg;
+                int readableBytes = buf.readableBytes();
+                stringBuffer.append(buf.readCharSequence(readableBytes, US_ASCII));
```

Comment:
wouldn't it be better to use a `ByteArrayOutputStream` and write into it ? The data may come in fragmented which means it may not be possible to read a valid charSequence... I guess it not matters to much for US_ASCII but I think the ByteArrayOutputStream is "more correct".